#6 6-refactor-dwcjinstaller-unzippom

kevinhagel opened the pull request:
- Created PomUnzipper.java
- created PomUnzipperTest.java
- created UnzipUtils.java
- created UnzipUtilsTest.java
- updated vavr version in pom
- Updated DwcjInstaller.java now using PomUnzipper.  Removed unzipPom method.
- Refactored DwcjInstaller install method.
- Refactored MavenOutputHandler.java now using PrintWriter, stops complaints from the IDE makes for cleaner code.

This is a major refactoring of DwcjInstall.  I will be creating two more such Issues for the other two internal methods.

Fix #9

hyyan commented:
@kevinhagel, the tests are failing due to an exception. Please ensure that the tests have everything necessary for direct execution.

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.dwcj.installer.util.UnzipUtilsTest
testJarFile = null
result      = Left(java.nio.file.NoSuchFileException: null)
[ERROR] Tests run: 10, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 0.071 s <<< FAILURE! -- in org.dwcj.installer.util.UnzipUtilsTest
[ERROR] org.dwcj.installer.util.UnzipUtilsTest.test_readyPomDestination_file_not_exist -- Time elapsed: 0.028 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
        at org.dwcj.installer.util.UnzipUtilsTest.getTestSource(UnzipUtilsTest.java:68)
        at org.dwcj.installer.util.UnzipUtilsTest.test_readyPomDestination_file_not_exist(UnzipUtilsTest.java:108)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] org.dwcj.installer.util.UnzipUtilsTest.test_readyPomDestination_mkdirs -- Time elapsed: 0.002 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
        at org.dwcj.installer.util.UnzipUtilsTest.getTestSource(UnzipUtilsTest.java:68)
        at org.dwcj.installer.util.UnzipUtilsTest.test_readyPomDestination_mkdirs(UnzipUtilsTest.java:83)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] org.dwcj.installer.util.UnzipUtilsTest.test_getChecksum_valid_file -- Time elapsed: 0.003 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
        at org.dwcj.installer.util.UnzipUtilsTest.getTestSource(UnzipUtilsTest.java:68)
        at org.dwcj.installer.util.UnzipUtilsTest.test_getChecksum_valid_file(UnzipUtilsTest.java:32)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] org.dwcj.installer.util.UnzipUtilsTest.test_readyPomDestination_no_mkdirs -- Time elapsed: 0.002 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
        at org.dwcj.installer.util.UnzipUtilsTest.getTestSource(UnzipUtilsTest.java:68)
        at org.dwcj.installer.util.UnzipUtilsTest.test_readyPomDestination_no_mkdirs(UnzipUtilsTest.java:73)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[INFO] Running org.dwcj.installer.PomUnzipperTest
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
[ERROR] Tests run: 4, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 0.554 s <<< FAILURE! -- in org.dwcj.installer.PomUnzipperTest
[ERROR] org.dwcj.installer.PomUnzipperTest.test_file_not_zip -- Time elapsed: 0.534 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
        at org.dwcj.installer.PomUnzipperTest.getTestSource(PomUnzipperTest.java:118)
        at org.dwcj.installer.PomUnzipperTest.test_file_not_zip(PomUnzipperTest.java:87)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] org.dwcj.installer.PomUnzipperTest.test_valid_zip_no_pom -- Time elapsed: 0.003 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
        at org.dwcj.installer.PomUnzipperTest.getTestSource(PomUnzipperTest.java:118)
        at org.dwcj.installer.PomUnzipperTest.test_valid_zip_no_pom(PomUnzipperTest.java:69)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[ERROR] org.dwcj.installer.PomUnzipperTest.test_unzip_valid_args -- Time elapsed: 0.002 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
        at org.dwcj.installer.PomUnzipperTest.getTestSource(PomUnzipperTest.java:118)
        at org.dwcj.installer.PomUnzipperTest.test_unzip_valid_args(PomUnzipperTest.java:54)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)

[INFO] Running org.dwcj.installer.PomParserTest
basedirPath =  /Users/kevin/bbx/plugins/DWCJ/deployments/DwcjSpringApplicationJava
nameCount = 8
foo = DwcjSpringApplicationJava
second foo =  /Users/kevin/bbx/plugins/DWCJ/deployments/DwcjSpringApplicationJava
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.062 s -- in org.dwcj.installer.PomParserTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   PomUnzipperTest.test_file_not_zip:87->getTestSource:118 NullPointer Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
[ERROR]   PomUnzipperTest.test_unzip_valid_args:54->getTestSource:118 NullPointer Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
[ERROR]   PomUnzipperTest.test_valid_zip_no_pom:69->getTestSource:118 NullPointer Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
[ERROR]   UnzipUtilsTest.test_getChecksum_valid_file:32->getTestSource:68 NullPointer Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
[ERROR]   UnzipUtilsTest.test_readyPomDestination_file_not_exist:108->getTestSource:68 NullPointer Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
[ERROR]   UnzipUtilsTest.test_readyPomDestination_mkdirs:83->getTestSource:68 NullPointer Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
[ERROR]   UnzipUtilsTest.test_readyPomDestination_no_mkdirs:73->getTestSource:68 NullPointer Cannot invoke "java.net.URL.getFile()" because the return value of "java.lang.ClassLoader.getResource(String)" is null
[INFO] 
[ERROR] Tests run: 19, Failures: 0, Errors: 7, Skipped: 0
```

hyyan commented:
@kevinhagel 

macOs Version 14.0
java 17

kevinhagel commented:
I am now trying a different way.  The maven build is not writing the resources into the target directory, and not seeing them on the classpath.  Using the classpath like this is standard in maven projects.  It is very curious it does not work here.
I have committed updates, using a different method

hyyan commented:
@kevinhagel Ok. let me know when this is running. Please not that `.gitignore` should ignore the target folder

kevinhagel commented:
added target/ to .gitignore
Can you tell me where the machine is that is running the maven build, so I can see the results

kevinhagel commented:
@hyyan Can you check it again?  The problem comes from the different way maven seems to run and the java.nio.Path runs on different systems.  Possibly also something that Juni5 does.
I converted the src/test/resources directory to an absolute path then resolved my test data against that.
I'm still very curious that the getClass().getClassLoader().getResourceAsStream wasn't working.
ALso -- the jar and zip resources I added to the repo were exccluded from being checked in by the .gitignore!  I fixed that.

hyyan commented on pom.xml:
```diff
@@ -31,11 +30,16 @@
-            <systemPath>${bbjlibdir}/BBjAdminAPI.jar</systemPath>
+            <groupId>com.basis.lib</groupId>
+            <artifactId>BBjAdminAPI</artifactId>
+            <version>23.06-SNAPSHOT</version>
```
It should be 23.06. no snapshots anymore for 23.06

hyyan commented on src/main/java/org/dwcj/installer/PomUnzipper.java:
```diff
@@ -0,0 +1,84 @@
+ * filesystem.
+ *
+ * @author Kevin Hagel
+ * @since 2024-01-24
```
[`@since` ](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#@since) Specify the product version when the class was added to the API specification. It should not be a date

hyyan commented on src/test/java/org/dwcj/installer/PomParserTest.java:
```diff
@@ -52,21 +54,25 @@ void test_valid_pom_no_configuration_children() {
     }
 
+
+    Path getTestSource(String testFileName) {
```
This method is repeated in differnt places. It should be moved to a utiltiy class

hyyan commented:
@kevinhagel looks good to me. I required few minor changes only

hyyan commented:
@MatthewHawkins please go over the docs and make sure the build is working

MatthewHawkins reviewed (changes requested):
Hey Kevin,

When I compiled this jar and used it with the install plugin, I had some issues when trying to deploy my demo samples.

The first deployment worked with the following output:

```
[INFO] -------DWCJ Deploy to Server:-------------
[INFO] Installing DWCJ App using URL: http://localhost:8888/dwcj-install
[INFO] dwcj-installer: received controlsamples.jar
install:
  sourceFilePaath = C:\bbx\cache\servletUploads\fileUpload935560162460111924.tmp
  jarFileName = controlsamples.jar
  bbxdir = C:\bbx
  deployroot = C:/bbx/plugins/DWCJ/deployments/
  requestUrl = http://localhost:8888
dwcj-install: WARNING: C:\bbx\plugins\DWCJ\deployments\controlsamples.jar was locked!
dwcj-installer: extracting pom.xml from zipFile C:\bbx\plugins\DWCJ\deployments\controlsamples.jar
dwcj-install: POM DESTINATION = C:\bbx\plugins\DWCJ\deployments\controlsamples\pom.xml
dwcj-installer: pom.xml not changed. No need to run Maven but running anyway.
dwcj-install: invocation request args = [-U, -DincludeScope=compile, -DoutputDirectory=C:\bbx\plugins\DWCJ\deployments\controlsamples\836244e2a65a4aca5bbab4f4bcc22d4f\target\dependency]
dwcj-install: Using Maven in location C:/bbx/plugins/DWCJ/deployments//apache-maven-3.8.8/bin/mvn
dwcj-install: INFO: org.apache.maven.shared.invoker.DefaultInvocationRequest@3f85c0d7
dwcj-installer: [INFO] Scanning for projects...
dwcj-installer: [INFO]
dwcj-installer: [INFO] ------------------< org.dwcj.samples:ControlSamples >-------------------
dwcj-installer: [INFO] Building ControlSamples 23.06
dwcj-installer: [INFO]   from pom.xml
dwcj-installer: [INFO] --------------------------------[ jar ]---------------------------------
dwcj-installer: [INFO]
dwcj-installer: [INFO] --- maven-dependency-plugin:2.8:copy-dependencies (default-cli) @ ControlSamples ---
dwcj-installer: [INFO] dwcj-engine-23.06.jar already exists in destination.
dwcj-installer: [INFO] dwcj-code-23.06.jar already exists in destination.
dwcj-installer: [INFO] gson-2.10.1.jar already exists in destination.
dwcj-installer: [INFO] ------------------------------------------------------------------------
dwcj-installer: [INFO] BUILD SUCCESS
dwcj-installer: [INFO] ------------------------------------------------------------------------
dwcj-installer: [INFO] Total time:  2.119 s
dwcj-installer: [INFO] Finished at: 2024-01-31T12:46:29+01:00
dwcj-installer: [INFO] ------------------------------------------------------------------------
dwcj-installer: Scanning Dependencies in C:\bbx\plugins\DWCJ\deployments\controlsamples\836244e2a65a4aca5bbab4f4bcc22d4f\target\dependency
dwcj-installer: extracting C:\bbx\plugins\DWCJ\deployments\controlsamples\bbj\DwcjHelper.bbj
dwcj-installer: extracting C:\bbx\plugins\DWCJ\deployments\controlsamples\bbj\dwcj.bbj
INFO: Written Classpath dwcj_controlsamples with 3 entries
INFO: Created App Deployment for  http://localhost:8888/webapp/controlsamples
```


However, when I changed my code and ran the install command again, I got the following output and no updated application:

```
[INFO] -------DWCJ Deploy to Server:-------------
[INFO] Installing DWCJ App using URL: http://localhost:8888/dwcj-install
[INFO] dwcj-installer: received controlsamples.jar
Error in installer core:
Error 26 Not a string: (null) on line 43 in Executed BBj Snippet
```

hyyan commented:
@kevinhagel Are you testing with hello world sample? Try with [ControlSamples](https://github.com/DwcJava/ControlSamples)

hyyan commented on pom.xml:
```diff
@@ -22,6 +22,11 @@
+        <dependency>
+            <groupId>log4j</groupId>
+            <artifactId>log4j</artifactId>
+            <version>1.2.17</version>
```
Update to [log4j2](https://logging.apache.org/log4j/2.x/)
